src/cpp/session/SessionOfflineService.cpp

```diff
@@ -0,0 +1,301 @@
+      }
+
+      return asyncConnection;
+   }
+   return boost::shared_ptr<HttpConnection>();
+}
+
+void OfflineService::run()
+{
+   try
+   {    
+      // default time durations
+      int asyncRpcMillis = options().asyncRpcTimeoutMs();
+      int handleOfflineMillis = options().handleOfflineTimeoutMs();
+      if (!options().asyncRpcEnabled())
+         asyncRpcMillis = 0;
+      if (!options().handleOfflineEnabled())
+         handleOfflineMillis = 0;
+
+      if (asyncRpcMillis != 0 && asyncRpcMillis < 100)
+      {
+         asyncRpcMillis = 100;
+      }
+      if (handleOfflineMillis != 0 && handleOfflineMillis < 200)
+      {
+         handleOfflineMillis = 200;
+      }
+
+      int sleepMillis = asyncRpcMillis;
+      if (sleepMillis == 0)
+         sleepMillis = 250;
+
+      // Pick the smallest non-zero value to use in the sleep call
+      // TODO: asyncRpc will run jobs later in some cases e.g. asyncRpcMillis=500 and handleOffline=750
+      // runs offline requests every second, not 750. Keeping it simple for now
+      if (sleepMillis == 0 || (handleOfflineMillis != 0 && handleOfflineMillis < sleepMillis))
+         sleepMillis = handleOfflineMillis;
+
+      s_asyncRpcDuration = std::chrono::milliseconds(asyncRpcMillis);
+      s_handleOfflineDuration = std::chrono::milliseconds(handleOfflineMillis);
+
+      LOG_DEBUG_MESSAGE("OfflineService started with session-async-rpc-timeout-ms=" + std::to_string(asyncRpcMillis) +
+                        " handle-offline-timeout-ms=" + std::to_string(handleOfflineMillis) + " polling every: " + std::to_string(sleepMillis) + "ms");
+
+      boost::chrono::milliseconds sleepTime = boost::chrono::milliseconds(sleepMillis);
+
+      // get alias to mainConnectionQueue
+      HttpConnectionQueue& mainConnectionQueue = httpConnectionListener().mainConnectionQueue();
+      bool serverStopped = false;
+      int emitMemEventsCt = prefs::userPrefs().memoryQueryIntervalSeconds() * 1000 / sleepMillis;
+      if (emitMemEventsCt == 0)
+         emitMemEventsCt = 1;
+      int emitMemEventsIndex = 0;
+      while (!serverStopped)
+      {
+         if (boost::this_thread::interruption_requested())
+         {
+            serverStopped = true;
+            break;
+         }
+         try
+         {
+            // Wake up using the minimum interval
+            boost::this_thread::sleep_for(sleepTime);
+
+            // Wait for the client to be initialized before any offline handling
+            if (!httpConnectionListener().eventsActive())
+               continue;
+
+            // If R is not occupying the main thread, we'll continue to wait.
+            // TODO: asyncRpc - are there other skippable operations, not from R - if so, we could remove this test
+            if (!console_input::executing())
+            {
+               // TODO: asyncRpc log a warning here if mainConnectionQueue has more than 5 entries
```

Comment:
As above -- do these TODOs need to be resolved before commit? If you intend to commit them, let's change them to "consider"